src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -6,19 +6,137 @@
  * За бажанням можна реалізувати інші методи інтрефейсу Map.</p>
  */
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int DEFAULT_CAPACITY = 16;
+    private static final double LOAD_FACTOR = 0.75;
+    private static final int RESIZE_RATE = 2;
+    private static final int INITIAL_SIZE = 0;
+    private Node<K, V>[] table;
+    private int size;
+    private int threshold;
+
+    public MyHashMap() {
+        table = new Node[DEFAULT_CAPACITY];
+        size = INITIAL_SIZE;
+        threshold = (int) (DEFAULT_CAPACITY * LOAD_FACTOR);
+    }
 
     @Override
     public void put(K key, V value) {
-
+        if (size == threshold) {
+            tableResize();
+        }
+        if (key == null) {
+            putForNullKey(value);
+            return;
+        }
+        int index = indexFor(hash(key.hashCode()), table.length);
+        addNode(key, value, index);
     }
 
     @Override
     public V getValue(K key) {
-        return null;
+        if (key == null) {
+            return getForNullKey();
+        }
+        Node<K, V> node = getNode(key);
+        return node != null ? node.value : null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private V getForNullKey() {
+        for (Node<K, V> e = table[0]; e != null; e = e.next) {
+            if (e.key == null) {
+                return e.value;
+            }
+        }
+        return null;
+    }
+
+    private Node<K, V> getNode(K key) {
+        for (Node<K, V> node : table) {
+            while (node != null) {
+                if (isEqual(node.key, key)) {
+                    return node;
+                }
+                node = node.next;
+            }
+        }
+        return null;
+    }
+
+    private void tableResize() {
+        Node<K, V>[] newTable = new Node[table.length * RESIZE_RATE];
+        transfer(newTable);
+        table = newTable;
+        threshold = (int) (table.length * LOAD_FACTOR);
+    }
+
+    private void transfer(Node<K, V>[] newTable) {
+        for (Node<K, V> node : table) {
+            while (node != null) {
+                Node<K, V> next = node.next;
+                int newIndex = indexFor(hash(node.key.hashCode()),
+                        newTable.length);
+                node.next = newTable[newIndex];
+                newTable[newIndex] = node;
+                node = next;
+            }
+        }
+    }
+
+    private void putForNullKey(V value) {
+        if (setValue(table[0], null, value)) {
+            return;
+        }
+        addNode(null, value, 0);
```

Comment:
let's modify our indexFor() so it will return 0 for key == null. This way we can get rid of this(and similars) methods

Reply:
Agree. A lot of redundant methods. Fixed.